apps/checkout-app/backend/payments/providers/adyen/index.ts

```diff
@@ -65,9 +74,146 @@ export const createAdyenPayment = async (
   return url;
 };
 
-export const verifyPayment = async (
+export const isNotificationDuplicate = async (
+  transactions: TransactionFragment[],
+  notificationItem: Types.notification.NotificationRequestItem
+) => {
+  const eventKeys = transactions
+    .map(({ events }) => events.map(createEventUniqueKey))
+    .flat();
+  const newEventKey = createEventUniqueKey({
+    reference: notificationItem.pspReference,
+    name: notificationItem.eventCode.toString(),
+  });
+
+  return eventKeys.includes(newEventKey);
+};
+
+export const getOrderId = async (
+  notification: Types.notification.NotificationRequestItem
+) => {
+  const { additionalData } = notification;
+  const paymentLinkId = additionalData?.paymentLinkId;
+
+  if (!paymentLinkId) {
+    return;
+  }
+
+  const { metadata } = await checkout.getPaymentLinks(paymentLinkId);
+
+  return metadata?.orderId;
+};
+
+export const getUpdatedTransactionData = (
+  transaction: TransactionFragment,
   notification: Types.notification.NotificationRequestItem
-): Promise<TransactionCreateMutationVariables | undefined> => {
+): TransactionUpdateMutationVariables => {
+  const {
+    eventCode,
+    amount,
+    pspReference,
+    originalReference,
+    operations,
+    success,
+  } = notification;
+
+  if (!originalReference) {
+    throw "originalReference does not exit on notification";
+  }
+
+  const getStatus = (): TransactionStatus => {
+    const failureStates = [
+      Types.notification.NotificationRequestItem.EventCodeEnum.CaptureFailed,
+      Types.notification.NotificationRequestItem.EventCodeEnum.RefundFailed,
+    ];
+
+    if (
+      failureStates.includes(eventCode) ||
+      success === Types.notification.NotificationRequestItem.SuccessEnum.False
+    ) {
+      return "FAILURE";
+    }
+
+    return "SUCCESS";
+  };
+
+  const getNewAmount = ():
+    | Pick<
+        TransactionUpdateInput,
+        "amountRefunded" | "amountAuthorized" | "amountCharged" | "amountVoided"
+      >
+    | undefined => {
+    if (
+      eventCode ===
+      Types.notification.NotificationRequestItem.EventCodeEnum.Refund
+    ) {
+      if (!amount.currency || !amount.value) {
+        throw "Amount not specified for a refund notification";
+      }
+      const refundAmount = getSaleorAmountFromAdyen(amount.value);
+
+      if (transaction.chargedAmount.amount !== 0) {
+        const chargedAmount = transaction.chargedAmount.amount - refundAmount;
+
+        if (chargedAmount < 0) {
+          throw "Amount after refund cannot be negative";
+        }
+
+        return {
+          amountCharged: {
+            amount: chargedAmount,
+            currency: amount.currency!,
+          },
+          amountRefunded: {
+            amount: refundAmount,
+            currency: amount.currency!,
+          },
+        };
+      } else if (transaction.authorizedAmount.amount !== 0) {
+        const chargedAmount =
```

Comment:
So this part is basically the same for charges and authorized amount, I think it could be moved to a function like `extractAmountData` that takes in either chargedAmount or authorizedAmount ?